{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# First - read the manuscript\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2951089/\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What Can BioRuby Do?\n",
    "\n",
    "\n",
    "* Sequence and feature (e.g. gene, exon, intron) creating/reading/writing\n",
    "* Sequence database retrieval (e.g. embl, genbank, swissprot…)\n",
    "* Sequence Alignment (many many types: fasta, clustal, phylip…,..,..,..,..,...,...)\n",
    "* Sequence and feature database creation/query\n",
    "* Sequence search/analysis (e.g. blast, ‘virtual restrictions’, & many more)\n",
    "* Sequence annotation\n",
    "* Sequence assembly\n",
    "* Physical mapping (e.g. clones, contigs, etc.)\n",
    "* Taxonomy representation and database query/retrieval\n",
    "* Bio-ontology manipulation/exploration (e.g. Gene Ontology)\n",
    "* Phenotype representation and query (e.g. OMIM)\n",
    "* Phylogeny analyses\n",
    "* more more more more more more\n",
    "\n",
    "\n",
    "# We will start from the command-line\n",
    "\n",
    "BioRuby provides its own command line (i.e. its own \"terminal window\")\n",
    "\n",
    "    /Lectures/$  bioruby              < NOTE WHERE WE START!  IN THE LECTURES FOLDER!!!!\n",
    "    Creating directory (/home/osboxes/.bioruby/shell/session) ... done\n",
    "    Creating directory (/home/osboxes/.bioruby/shell/plugin) ... done\n",
    "    Creating directory (/home/osboxes/.bioruby/data) ... done\n",
    "\n",
    "    . . . B i o R u b y   i n   t h e   s h e l l . . .\n",
    "\n",
    "      Version : BioRuby 1.5.1 / Ruby 2.4.2\n",
    "\n",
    "    bioruby> \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demo of FASTQ Sequence quality masking\n",
    "\n",
    "It is important to mask low-quality base reads before doing a wide range of downstream operations like alignment or SNP-calling.\n",
    "\n",
    "We will now take a sequence file (in <code> Lectures/files/SP1.fq </code>) and we will use the BioRuby command line to mask it.\n",
    "\n",
    "The code to do this is:\n",
    "\n",
    "---------------------------\n",
    "    Bio::FlatFile.open('./files/SP1.fq').each do |entry|  # each sequence will be put into 'entry'\n",
    "        hq = entry.mask(20)                               # set the quality filter to 20\n",
    "        puts hq.output_fasta(entry.entry_id)              # print the FASTA out again, with the same ID as the head\n",
    "    end\n",
    "\n",
    "-----------------\n",
    "In BioRuby's command line it looks like this:\n",
    "\n",
    "\n",
    "<code>\n",
    "bioruby> Bio::FlatFile.open('./files/SP1.fq').each do |entry|\n",
    "**bioruby+**   hq = entry.mask(20)               < NOTE:  'bioruby+' means that it is waiting for more input\n",
    "puts hq.output_fasta(entry.entry_id)\n",
    "end\n",
    "<code>    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Command line as a debugger\n",
    "\n",
    "One possible use of the BioRuby command line is as a debugger - you can explore an object \"live\" until you decide what command you want to use, then you put that command into your code.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BioRuby - Documentation\n",
    "\n",
    "http://bioruby.org/rdoc/\n",
    "\n",
    "# Sequences and Sequence Features\n",
    "\n",
    "We will focus on Sequences in this class.  \n",
    "* how to retrieve them\n",
    "* how to write them to files\n",
    "* how to change their formats\n",
    "* how to explore their features (e.g. genes)\n",
    "* how to create databases from them to:\n",
    "* make them easier to find/query\n",
    "\n",
    "# BioRuby's Sequence objects\n",
    "\n",
    "Like all other Ruby objects, you create a new Sequence object using \"new\"\n",
    "\n",
    "First, look at the documentation for Bio::Sequence, Bio::Sequence::AA, Bio::Sequence::NA (http://bioruby.org/rdoc/Bio/Sequence.html and find the other two yourself :-) )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "require 'bio'\n",
    "\n",
    "puts \"MYSEQ0 = Bio::Sequence.new('ACTTTGC')\"\n",
    "myseq0 = Bio::Sequence.new(\"ACTTTGC\")\n",
    "puts myseq0.class\n",
    "puts myseq0.seq\n",
    "puts myseq0.seq.class\n",
    "puts\n",
    "\n",
    "puts \"MYSEQ1 = Bio::Sequence.auto('ACTTTGC')\"\n",
    "myseq1 = Bio::Sequence.auto(\"ACTTTGC\")\n",
    "puts myseq1.class\n",
    "puts myseq1.seq\n",
    "puts myseq1.seq.class\n",
    "puts\n",
    "\n",
    "puts \"MYSEQ2 = Bio::Sequence::NA.new('ACTTTGC')\"\n",
    "myseq2 = Bio::Sequence::NA.new(\"ACTTTGC\")\n",
    "puts myseq2.class\n",
    "puts myseq2.seq\n",
    "puts myseq2.seq.class\n",
    "puts\n",
    "\n",
    "#puts myseq2.public_methods.join(\"\\n\")\n",
    "puts \"myseq2 = #{myseq2}\"\n",
    "\n",
    "myseq3 = myseq2.reverse_complement\n",
    "puts \"myseq3 = #{myseq3}  (reverse complement)\"\n",
    "\n",
    "myseq2.reverse_complement!    #   <---- note the ! ->  When you see this in Ruby, it means the method will change the object\n",
    "puts \"myseq2 = #{myseq2} (reverse complement!)\"\n",
    "\n",
    "# this is the better way to do it:\n",
    "puts myseq2.to_s    # to_s means \"to string\" - you will see this in a lot of Ruby objects\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Limitations\n",
    "\n",
    "* BioRuby does not have built-in connections to every database on earth (e.g. no connection to TAIR or AraPort)\n",
    " * You will still need to use REST interfaces and regular expressions...often!\n",
    "* BioRuby does not recognize every kind of sequence identifier (e.g. At1g287748 means nothing to BioRuby)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Object Methods\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "aaseq = Bio::Sequence.auto(\"BBSTD\")\n",
    "puts aaseq.seq.class\n",
    "puts aaseq.seq.public_methods.sort\n",
    "puts\n",
    "aaseq = Bio::Sequence.auto(\"AAAACCCCCCCTTTTTTTGGGGGG\")\n",
    "puts aaseq.seq.class\n",
    "puts aaseq.seq.public_methods.sort\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## writing sequences\n",
    "\n",
    "* :output \n",
    "* :list_output_formats\n",
    "* :output_fasta\n",
    "\n",
    "Note that the documentation for Bio::Sequence says:\n",
    "\n",
    "    Included Modules\n",
    "\n",
    "        Bio::Sequence::Format\n",
    "        Bio::Sequence::SequenceMasker \n",
    "\n",
    "That's where these \"output\" methods are coming from...\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "aaseq.list_output_formats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "puts aaseq.output_fasta    # ask for fasta\n",
    "puts aaseq.output    # the default is fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "puts aaseq.output(:fastq)   # specify that you want fastq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## searching sequences \n",
    "\n",
    "see:  https://www.bioinformatics.org/sms/iupac.html\n",
    "\n",
    "This is accomplished in bioruby using the .to_re\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seq = Bio::Sequence::NA.new(\"actgggggatccc\")\n",
    "search = Bio::Sequence::NA.new(\"ATMM\")   # M indicates it is an A or C\n",
    "\n",
    "re = Regexp.new(search.to_re)\n",
    "puts re.source   # to see the content of a regualr expression call .source\n",
    "\n",
    "match = seq.seq.match(re)\n",
    "puts match\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other common sequence methods\n",
    "(from the BioRuby Tutorial)\n",
    "\n",
    "    seq = Bio::Sequence::NA.new(\"atgcatgcaaaa\")\n",
    "    puts seq.complement\n",
    "\n",
    "    puts seq.subseq(3,8) # gets subsequence of positions 3 to 8 (starting from 1)\n",
    "    puts seq.gc_percent \n",
    "    puts seq.composition \n",
    "    puts seq.translate \n",
    "    puts seq.translate(2)        # translate from frame 2\n",
    "    puts seq.translate(1,11)     # codon table 11\n",
    "    puts seq.translate.codes\n",
    "    puts seq.translate.names\n",
    "    puts seq.translate.composition\n",
    "    puts seq.translate.molecular_weight\n",
    "    puts seq.complement.translate\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seq = Bio::Sequence::NA.new(\"atgcatgcaaaa\")\n",
    "puts seq.complement\n",
    "\n",
    "puts seq.subseq(3,8) # gets subsequence of positions 3 to 8 (starting from 1)\n",
    "puts seq.gc_percent \n",
    "puts seq.composition \n",
    "puts seq.translate \n",
    "puts seq.translate(2)        # translate from frame 2\n",
    "puts seq.translate(1,11)     # codon table 11\n",
    "puts seq.translate.codes\n",
    "puts seq.translate.names\n",
    "puts seq.translate.composition\n",
    "puts seq.translate.molecular_weight\n",
    "puts seq.complement.translate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading common database formats\n",
    "\n",
    "BioRuby can import a wide range of Sequence files, in many formats:\n",
    "\n",
    "    $ ls -l /home/osboxes/.rvm/gems/ruby-2.4.2/gems/bio-1.5.1/lib/bio/db\n",
    "    total 324\n",
    "    -rw-r--r-- 1 osboxes osboxes  7198 Oct 26 05:58 aaindex.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 biosql\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 embl\n",
    "    -rw-r--r-- 1 osboxes osboxes 15444 Oct 26 05:58 fantom.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 fasta\n",
    "    -rw-r--r-- 1 osboxes osboxes  9384 Oct 26 05:58 fasta.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 fastq\n",
    "    -rw-r--r-- 1 osboxes osboxes 18543 Oct 26 05:58 fastq.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 genbank\n",
    "    -rw-r--r-- 1 osboxes osboxes 61643 Oct 26 05:58 gff.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 10036 Oct 26 05:58 go.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 kegg\n",
    "    -rw-r--r-- 1 osboxes osboxes  9021 Oct 26 05:58 lasergene.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  1478 Oct 26 05:58 litdb.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  7622 Oct 26 05:58 medline.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  5328 Oct 26 05:58 nbrf.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 12178 Oct 26 05:58 newick.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 56187 Oct 26 05:58 nexus.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 pdb\n",
    "    -rw-r--r-- 1 osboxes osboxes   517 Oct 26 05:58 pdb.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 phyloxml\n",
    "    -rw-r--r-- 1 osboxes osboxes 11233 Oct 26 05:58 prosite.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 14132 Oct 26 05:58 rebase.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 sanger_chromatogram\n",
    "    -rw-r--r-- 1 osboxes osboxes 14249 Oct 26 05:58 soft.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  6022 Oct 26 05:58 transfac.rb\n",
    "    \n",
    "(NOTE: some of these are folders, if the database contains different things - e.g. kegg has genes, pathways, compounds, etc.)\n",
    "\n",
    "BioRuby is also able to guess (usually correctly!) what kind of sequence file you give it.  \n",
    "\n",
    "So, you can either tell it what kind of data you have:\n",
    "\n",
    "---------------\n",
    "<code>\n",
    "    ff = Bio::FlatFile.new(Bio::GenBank, data)\n",
    "</code>\n",
    "----------------\n",
    "\n",
    "Or, you can ask BioRuby to look at the data and make the decision itself:\n",
    "\n",
    "----------------------\n",
    "<code>\n",
    "    ff = Bio::FlatFile.auto(data)\n",
    "\n",
    "    ff.each_entry do |entry|    # many database formats, like FASTA, allow multiple records in a single file/string\n",
    "      p entry.entry_id          # identifier of the entry\n",
    "      p entry.definition        # definition of the entry\n",
    "      p entry.seq               # sequence data of the entry\n",
    "    end\n",
    "</code>\n",
    "----------------------\n",
    "\n",
    "## Let's try this for ourselves:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D   3    standard; DNA; HTG; 2175 BP.\n",
      "XX\n",
      "AC   chromosome:TAIR10:3:20119140:20121314:1\n",
      "XX\n",
      "SV   chromo\n",
      "Trying Method 1a - directly create a Bio::EMBL object from string\n",
      "Bio::EMBL\n",
      "The record is Arabidopsis thaliana chromosome 3 TAIR10 partial sequence 20119140..20121314 annotated by Araport \n",
      "\n",
      "\n",
      "Trying method 2 - create a new FlatFile object of the correct type\n",
      "Bio::EMBL\n",
      "Bio::FlatFile\n",
      "Bio::EMBL\n",
      "The record is Arabidopsis thaliana chromosome 3 TAIR10 partial sequence 20119140..20121314 annotated by Araport \n",
      "Bio::EMBL\n",
      "The record is  \n",
      "\n",
      "\n",
      "Trying method 3 - create a new FlatFile object using auto-detect for type\n",
      "Bio::FlatFile\n",
      "Bio::EMBL\n",
      "The record is *Arabidopsis thaliana chromosome 3 TAIR10 partial sequence 20119140..20121314 annotated by Araport* \n",
      "Bio::EMBL\n"
     ]
    }
   ],
   "source": [
    "require 'net/http'   # this is how you access the Web\n",
    "require 'bio'\n",
    "address = URI('http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=At3g54340')  # create a \"URI\" object (Uniform Resource Identifier: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier)\n",
    "\n",
    "response = Net::HTTP.get_response(address)  # use the Net::HTTP object \"get_response\" method\n",
    "                                             \n",
    "record = response.body\n",
    "puts record[1..100]  # show just a little bit...\n",
    "\n",
    "# create a local file with this data\n",
    "File.open('At3g54340.embl', 'w') do |myfile|  # w makes it writable\n",
    "  myfile.puts record\n",
    "end\n",
    "\n",
    "\n",
    "puts \"Trying Method 1a - directly create a Bio::EMBL object from string\"\n",
    "#  METHOD 1a - Create an object of the correct type from a string\n",
    "entry = Bio::EMBL.new(record)\n",
    "puts entry.class\n",
    "puts \"The record is #{entry.definition} \"\n",
    "\n",
    "\n",
    "\n",
    "#puts \"\\n\\nTrying Method 1b - directly create a Bio::EMBL object from a file\"\n",
    "# METHOD 1b - Create an object of the correct type from a file\n",
    "#entry = Bio::EMBL.open('At3g54340.embl')\n",
    "#puts entry.class\n",
    "#puts \"The record is #{entry.definition} \"\n",
    "\n",
    "\n",
    "# ==========================================================\n",
    "# see documentation at http://bioruby.org/rdoc/Bio/FlatFile\n",
    "\n",
    "puts \"\\n\\nTrying method 2 - create a new FlatFile object of the correct type\"\n",
    "\n",
    "# we can ask what type it is...\n",
    "datafiletype = Bio::FlatFile.autodetect(record)\n",
    "puts datafiletype\n",
    "\n",
    "datafile1 = Bio::FlatFile.new(datafiletype, File.open('At3g54340.embl', 'r')) # use that to create the correct type\n",
    "# note that the .new requires a \"stream\" as its second argument.  This is ugly, but not difficult...\n",
    "\n",
    "puts datafile1.class \n",
    "\n",
    "datafile1.each_entry do |entry|   # the FILE is not the same as the RECORD - multiple records can exist in a file\n",
    "  puts entry.class\n",
    "  puts \"The record is #{entry.definition} \"\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ==========================================================\n",
    "# see documentation at http://bioruby.org/rdoc/Bio/FlatFile\n",
    "\n",
    "puts \"\\n\\nTrying method 3 - create a new FlatFile object using auto-detect for type\"\n",
    "datafile2 = Bio::FlatFile.auto('At3g54340.embl')\n",
    "puts datafile2.class  \n",
    "\n",
    "datafile2.each_entry do |entry| # the FILE is not the same as the RECORD - multiple records can exist in a file\n",
    "  puts entry.class\n",
    "  puts \"The record is *#{entry.definition}* \" unless entry.definition.nil? || entry.definition.empty?  # see this? Useful!\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Methods available for the common Sequence files:\n",
    "\n",
    "http://bioruby.org/rdoc/Bio/EMBL\n",
    "\n",
    "\n",
    "    #cc\n",
    "    #comment\n",
    "    #data_class\n",
    "    #date_created\n",
    "    #date_modified\n",
    "    #dblinks\n",
    "    #division\n",
    "    #dt\n",
    "    #each_cds\n",
    "    #each_gene\n",
    "    #entry  *******\n",
    "    #entry_id\n",
    "    #entry_name\n",
    "    #entry_version\n",
    "    #features\n",
    "    #fh\n",
    "    #ft\n",
    "    #id_line\n",
    "    #molecule\n",
    "    #molecule_type\n",
    "    #naseq\n",
    "    #ntseq\n",
    "    #os\n",
    "    #release_created\n",
    "    #release_modified\n",
    "    #seq\n",
    "    #seqlen\n",
    "    #sequence_length\n",
    "    #species    ************\n",
    "    #sq\n",
    "    #sv\n",
    "    #to_biosequence\n",
    "    #topology\n",
    "    #version \n",
    "    \n",
    "    http://bioruby.org/rdoc/Bio/EMBLDB/Common.html\n",
    "    \n",
    "    #ac\n",
    "    #accession   *********\n",
    "    #accessions\n",
    "    #de\n",
    "    #definition\n",
    "    #description\n",
    "    #dr\n",
    "    #keywords\n",
    "    #kw\n",
    "    #oc\n",
    "    #og\n",
    "    #os\n",
    "    #ref\n",
    "    #references \n",
    "    \n",
    "    http://bioruby.org/rdoc/Bio/DB.html\n",
    "    \n",
    "    #entry_id\n",
    "    #exists?\n",
    "    #fetch\n",
    "    #get\n",
    "    #tags "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence features - exons, introns, polyA sites, etc.\n",
    "\n",
    "http://bioruby.org/rdoc/Bio/Feature.html\n",
    "\n",
    "http://bioruby.org/rdoc/Bio/Feature/Qualifier.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bio::FlatFile\n",
      "Bio::EMBL\n",
      "# chromosome:TAIR10:3:20119140:20121314:1 - Arabidopsis thaliana (thale-cress)\n",
      "Bio::Feature\n",
      "source\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1..2175\n",
      "Associations = {\"organism\"=>\"Arabidopsis thaliana\", \"db_xref\"=>\"taxon:3702\"}\n",
      "Bio::Feature\n",
      "gene\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1..2175)\n",
      "Associations = {\"gene\"=>\"AT3G54340\", \"locus_tag\"=>\"AP3\", \"note\"=>\"Floral homeotic protein APETALA 3 [Source:UniProtKB/Swiss-Prot;Acc:P35632]\"}\n",
      "Bio::Feature\n",
      "mRNA\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = join(complement(1761..2175),complement(1593..1659),complement(1448..1509),complement(1264..1363),complement(1050..1091),complement(729..773),complement(1..483))\n",
      "Associations = {\"gene\"=>\"AT3G54340\", \"standard_name\"=>\"AT3G54340.1\"}\n",
      "Bio::Feature\n",
      "CDS\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = join(complement(1761..1948),complement(1593..1659),complement(1448..1509),complement(1264..1363),complement(1050..1091),complement(729..773),complement(289..483))\n",
      "Associations = {\"gene\"=>\"AT3G54340\", \"protein_id\"=>\"AT3G54340.1.1\", \"note\"=>\"transcript_id=AT3G54340.1\", \"db_xref\"=>\"UniParc:UPI0000001796\", \"translation\"=>\"MARGKIQIKRIENQTNRQVTYSKRRNGLFKKAHELTVLCDARVSIIMFSSSNKLHEYISPNTTTKEIVDLYQTISDVDVWATQYERMQETKRKLLETNRNLRTQIKQRLGECLDELDIQELRRLEDEMENTFKLVRERKFKSLGNQIETTKKKNKSQQDIQKNLIHELELRAEDPHYGLVDNGGDYDSVLGYQIEGSRAYALRFHQNHHHYYPNHGLHAPSASDIITFHLLE\"}\n",
      "TRANSCRIPT FOUND!\n",
      "Gene Info:  AT3G54340, transcript_id=AT3G54340.1\n",
      "\n",
      "\n",
      ">NA splicing('join(complement(1761..1948),complement(1593..1659),complement(1448..1509),complement(1264..1363),complement(1050..1091),complement(729..773),complement(289..483))') : AT3G54340, transcript_id=AT3G54340.1\n",
      "Bio::Sequence::NA\n",
      "atggcgagagggaagatccagatcaagaggatagagaaccagacaaacagacaagtgacgtattcaaagagaagaaatggtttattcaagaaagcacatgagctcacggttttgtgtgatgctagggtttcgattatcatgttctctagctccaacaagcttcatgagtatatcagccctaacaccacaacgaaggagatcgtagatctgtaccaaactatttctgatgtcgatgtttgggccactcaatatgagcgaatgcaagaaaccaagaggaaactgttggagacaaatagaaatctccggactcagatcaagcagaggctaggtgagtgtttggacgagcttgacattcaggagctgcgtcgtcttgaggatgaaatggaaaacactttcaaactcgttcgcgagcgcaagttcaaatctcttgggaatcagatcgagaccaccaagaaaaagaacaaaagtcaacaagacatacaaaagaatctcatacatgagctggaactaagagctgaagatcctcactatggactagtagacaatggaggagattacgactcagttcttggataccaaatcgaagggtcacgtgcttacgctcttcgtttccaccagaaccatcaccactattaccccaaccatggccttcatgcaccctctgcctctgacatcattaccttccatcttcttgaataa\n",
      "\n",
      "\n",
      ">AA translated by splicing('join(complement(1761..1948),complement(1593..1659),complement(1448..1509),complement(1264..1363),complement(1050..1091),complement(729..773),complement(289..483))').translate\n",
      "MARGKIQIKRIENQTNRQVTYSKRRNGLFKKAHELTVLCDARVSIIMFSSSNKLHEYISPNTTTKEIVDLYQTISDVDVWATQYERMQETKRKLLETNRNLRTQIKQRLGECLDELDIQELRRLEDEMENTFKLVRERKFKSLGNQIETTKKKNKSQQDIQKNLIHELELRAEDPHYGLVDNGGDYDSVLGYQIEGSRAYALRFHQNHHHYYPNHGLHAPSASDIITFHLLE*\n",
      "\n",
      "\n",
      ">AA original translation\n",
      "MARGKIQIKRIENQTNRQVTYSKRRNGLFKKAHELTVLCDARVSIIMFSSSNKLHEYISPNTTTKEIVDLYQTISDVDVWATQYERMQETKRKLLETNRNLRTQIKQRLGECLDELDIQELRRLEDEMENTFKLVRERKFKSLGNQIETTKKKNKSQQDIQKNLIHELELRAEDPHYGLVDNGGDYDSVLGYQIEGSRAYALRFHQNHHHYYPNHGLHAPSASDIITFHLLE\n",
      "Bio::Feature\n",
      "exon\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1..483)\n",
      "Associations = {\"note\"=>\"exon_id=AT3G54340.1.exon7\"}\n",
      "Bio::Feature\n",
      "exon\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1050..1091)\n",
      "Associations = {\"note\"=>\"exon_id=AT3G54340.1.exon5\"}\n",
      "Bio::Feature\n",
      "exon\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(729..773)\n",
      "Associations = {\"note\"=>\"exon_id=AT3G54340.1.exon6\"}\n",
      "Bio::Feature\n",
      "exon\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1593..1659)\n",
      "Associations = {\"note\"=>\"exon_id=AT3G54340.1.exon2\"}\n",
      "Bio::Feature\n",
      "exon\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1448..1509)\n",
      "Associations = {\"note\"=>\"exon_id=AT3G54340.1.exon3\"}\n",
      "Bio::Feature\n",
      "exon\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1761..2175)\n",
      "Associations = {\"note\"=>\"exon_id=AT3G54340.1.exon1\"}\n",
      "Bio::Feature\n",
      "exon\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1264..1363)\n",
      "Associations = {\"note\"=>\"exon_id=AT3G54340.1.exon4\"}\n",
      "Bio::Feature\n",
      "mRNA\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = join(complement(AL132971.2:19369..19757),complement(AL132971.2:19047..19281),complement(AL132971.2:18782..18927),complement(AL132971.2:18060..18147),complement(AL132971.2:17832..17978),complement(AL132971.2:16910..17059),complement(AL132971.2:14556..14644),complement(AL132971.2:14227..14290),complement(AL132971.2:13793..13853),complement(AL132971.2:13177..13184),complement(AL132971.2:11590..13131),complement(AL132971.2:11453..11504),complement(AL132971.2:11074..11222),complement(AL132971.2:10884..10970),complement(AL132971.2:10633..10761),complement(1448..1509),complement(1264..1363),complement(1050..1091),complement(729..773),complement(289..483))\n",
      "Associations = {\"product\"=>\"MTTTSVCPFSKAARPDDGSTRKQGEITASGCPFSKAARPDDASARKQGE TTASGCPFSKSARPDENGSKGCPEQEGNLNKDSTDSATVPAKCPFGYDSQTFKLGPFSC MLCQALLYESSRCVPCTHVFCKVCLTRFKDCPLCGADIESIEVDENLQKMVDQFIEGHA RIKRSVVNGTEKEEVENDNKKVIYADVSMERGSFLVQQAMRAFSAQNYESAKSRLAMCT EDIRDQLGREGNTPELCSQLGAVLGMLGDCSRAMGDSSSAVKHFEESVEFLMKLPLNDL EITHTLSVSLNKIGDLKYYDEDLQAARSYYDRALNVRRDAMKHHPNAPSQILDVAVSLA KVADIDRTLQNEVAATDGFKEGMRLLESLKLDSEDSALEQRFCDMVTVARLLNLTLVVP EIDKTSFFADSRFCVLSPSLRPLLRYAFLREVKLLVSIVDFEGIGEYDISFEAACLAGA SLESLAKGAVQFSRRFSIRELQDRWHALLYDPVVSVEAAFRMAELERTNPNFPTKFGRT GYSKENKSSSRKRNAERLRSTYHSLRKKFRTEPFNSLDLGFLVPPNDSHFMDNGDATHL GLEDSHMDIIHNAFPEILAEGGCVTTHVLPEDNLQGDIPYVEGENLTFTEHAGLSVCDV VHQDSEQKLENTAHEAKNTMASTDFLAQLSTSLFEEDMEPFMEVDGKEVDKSYYDGLSS LLVNSTNDTNREAFPNPTEQEPSIAPTHPGEATLDDHVMLELDGTIALDPHPEIVGGVI CCLLNEEDPDIPCNDDIFLSNNSRPMSVSSLARRNFKDTNSPITTCVRDVSASKEKSEG YSLQAQKKKPGRLQGSTQGKPEMGQPSKGSKFRASTSTELKNTVAPGGSSSAQACSNTL LSTGTGAKDGKKETATGTLFVGSDGHGNHPEKDSENCKEKNVVPPVNESPHAKDTDDGL IEITVPELEITRAEAEAEAEAHVCESDEDLPNYSDIEAMILDMDLEPDDQDNFDLEVSK YQSQDMKRTIIRLEQAAHSYMQRAIASRGAFAVLYGRYSKHYIKKPEVLVGRSTEDLAV DIDLGREKRGSKISRRQAIIRLGDDGSFHIKNLGKYSISVNEKEVDPGQSLILKSDCLV ERMQETKRKLLETNRNLRTQIKQRLGECLDELDIQELRRLEDEMENTFKLVRERKFKSL GNQIETTKKKNKSQQDIQKNLIHELELRAEDPHYGLVDNGGDYDSVLGYQIEGSRAYAL RFHQNHHHYYPNHGLHAPSASDIITFHLLE\", \"note\"=>\"Derived by automated computational analysis using gene prediction method:fgenesh\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 18..18\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313434\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 22..22\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503635\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 23..23\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119162_A_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 38..38\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313435\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 40..40\n",
      "Associations = {\"replace\"=>\"A/-\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH15197512\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 52..52\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119191_T_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 68..68\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503636\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 73..73\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH00418964\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 89..89\n",
      "Associations = {\"replace\"=>\"T/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119228_T_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 102..102\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432852\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 145..145\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119284_C_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 179..179\n",
      "Associations = {\"replace\"=>\"C/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119318_C_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 185..185\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313438\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 189..189\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503637\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 195..194\n",
      "Associations = {\"replace\"=>\"-/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119333_G_GT\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 210..210\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313439\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 225..224\n",
      "Associations = {\"replace\"=>\"-/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119363_G_GT\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 249..249\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119388_G_A\"}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 264..264\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119403_C_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 333..333\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH00418965\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 352..352\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503638\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 357..357\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503639\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 380..380\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503640\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 385..385\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755904\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 400..400\n",
      "Associations = {\"replace\"=>\"C/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119539_C_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 454..454\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119593_T_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 455..455\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119594_C_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 502..501\n",
      "Associations = {\"replace\"=>\"-/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119640_G_GT\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 537..537\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119676_T_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 590..590\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755925\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 603..603\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313441\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 624..624\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH00418966\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 627..627\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119766_T_C\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 654..654\n",
      "Associations = {\"replace\"=>\"T/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119793_T_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 700..700\n",
      "Associations = {\"replace\"=>\"C/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313442\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 708..708\n",
      "Associations = {\"replace\"=>\"A/-\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119846_GA_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 760..760\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755926\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 786..786\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503641\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 792..792\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432854\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 802..802\n",
      "Associations = {\"replace\"=>\"G/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503642\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 819..819\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503643\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 820..820\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755929\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 822..822\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119961_C_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 825..825\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119964_A_C\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 845..844\n",
      "Associations = {\"replace\"=>\"-/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH04251560\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 848..847\n",
      "Associations = {\"replace\"=>\"-/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119986_C_CA\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 857..857\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20119996_T_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 867..867\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120006_A_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 876..876\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755932\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 884..885\n",
      "Associations = {\"replace\"=>\"TT/-\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH15365420\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 906..909\n",
      "Associations = {\"replace\"=>\"ACAC/AC/ACACAC/-\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH15365421\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 915..915\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120054_C_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 926..926\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120065_A_C\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 980..980\n",
      "Associations = {\"replace\"=>\"A/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313444\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 984..984\n",
      "Associations = {\"replace\"=>\"A/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH00418967\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1063..1063\n",
      "Associations = {\"replace\"=>\"G/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH00418968\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1119..1119\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120258_A_C\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1140..1140\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432875\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1144..1144\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120283_T_C\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1162..1162\n",
      "Associations = {\"replace\"=>\"A/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313446\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1174..1174\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503644\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1183..1183\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503645\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1185..1185\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503646\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1186..1186\n",
      "Associations = {\"replace\"=>\"A/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503647\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1188..1193\n",
      "Associations = {\"replace\"=>\"TAATTC/GAATTC/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120327_TAATTC_GAATTC,T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1198..1198\n",
      "Associations = {\"replace\"=>\"C/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755933\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1209..1209\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120348_C_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1223..1223\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120362_A_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1239..1239\n",
      "Associations = {\"replace\"=>\"G/-\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120377_AG_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1275..1275\n",
      "Associations = {\"replace\"=>\"G/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755934\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1276..1276\n",
      "Associations = {\"replace\"=>\"A/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503648\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1289..1289\n",
      "Associations = {\"replace\"=>\"G/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120428_G_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1318..1318\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120457_C_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1320..1320\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755935\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1323..1323\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120462_C_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1329..1329\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313448\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1338..1338\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503649\"}\n",
      "Bio::Feature\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1363..1363\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432876\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1374..1374\n",
      "Associations = {\"replace\"=>\"A/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755936\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1382..1382\n",
      "Associations = {\"replace\"=>\"G/C/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755937\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1390..1390\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432877\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1395..1395\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120534_C_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1408..1408\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313449\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1441..1441\n",
      "Associations = {\"replace\"=>\"C/A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313450\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1450..1450\n",
      "Associations = {\"replace\"=>\"G/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120589_G_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1453..1453\n",
      "Associations = {\"replace\"=>\"C/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755938\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1571..1571\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120710_A_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1585..1585\n",
      "Associations = {\"replace\"=>\"A/-\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120723_GA_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1586..1586\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313451\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1628..1628\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755939\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1647..1647\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755940\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1670..1670\n",
      "Associations = {\"replace\"=>\"C/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503650\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1675..1675\n",
      "Associations = {\"replace\"=>\"T/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120814_T_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1685..1685\n",
      "Associations = {\"replace\"=>\"A/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432878\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1704..1704\n",
      "Associations = {\"replace\"=>\"A/-\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120842_GA_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1712..1712\n",
      "Associations = {\"replace\"=>\"G/C\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120851_G_C\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1735..1735\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120874_G_A\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1747..1747\n",
      "Associations = {\"replace\"=>\"G/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20120886_G_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1811..1811\n",
      "Associations = {\"replace\"=>\"G/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503651\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1868..1868\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121007_A_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1874..1874\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503652\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1889..1889\n",
      "Associations = {\"replace\"=>\"C/T\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121028_C_T\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1916..1916\n",
      "Associations = {\"replace\"=>\"T/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121055_T_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1954..1954\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432879\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1980..1980\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121119_A_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1987..1987\n",
      "Associations = {\"replace\"=>\"T/C\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121126_T_C\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2000..2000\n",
      "Associations = {\"replace\"=>\"C/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH12755941\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2012..2012\n",
      "Associations = {\"replace\"=>\"T/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121151_T_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2016..2016\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121155_A_G\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2131..2131\n",
      "Associations = {\"replace\"=>\"G/C\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH14432880\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2138..2138\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503653\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2145..2144\n",
      "Associations = {\"replace\"=>\"-/A\", \"db_xref\"=>\"The 1001 Genomes Project:tmp_3_20121283_T_TA\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2165..2165\n",
      "Associations = {\"replace\"=>\"C/G/T\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503654\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2168..2168\n",
      "Associations = {\"replace\"=>\"A/G\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH02503655\"}\n",
      "Bio::Feature\n",
      "variation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 2169..2169\n",
      "Associations = {\"replace\"=>\"T/A\", \"db_xref\"=>\"The 1001 Genomes Project:ENSVATH06313453\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 36..1979\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-171595 : 3..985(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 227..1978\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-171708 : 1..486(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 111..428\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-298790 : 1..319(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 448..1991\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-5380411836 : 2..584(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 34..2031\n",
      "Associations = {\"note\"=>\"match: At.287 : 1..1037(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 289..1948\n",
      "Associations = {\"note\"=>\"match: TC293195 : 104..802(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 62..1978\n",
      "Associations = {\"note\"=>\"match: TC301920 : 3..958(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1078..1957\n",
      "Associations = {\"note\"=>\"match: TC305197 : 244..677(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 227..1978\n",
      "Associations = {\"note\"=>\"match: TC305959 : 1..486(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1268..2031\n",
      "Associations = {\"note\"=>\"match: TC306275 : 1..496(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 34..1978\n",
      "Associations = {\"note\"=>\"match: TA29636_3702 : 2..985(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 227..1978\n",
      "Associations = {\"note\"=>\"match: TA29642_3702 : 1..486(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = AL132971.2:7054..8774\n",
      "Associations = {\"note\"=>\"match: TA29643_3702 : 2..627(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1848..1903\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Medicago_truncatula-44769 : 89..144(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1848..1903\n",
      "Associations = {\"note\"=>\"match: Mtr.16663 : 58..113(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1848..1903\n",
      "Associations = {\"note\"=>\"match: TC110754 : 89..144(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1848..1903\n",
      "Associations = {\"note\"=>\"match: TA32762_3880 : 89..144(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1779..1941\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Arachis_hypogaea-2514 : 224..386(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1825..1939\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Gossypium-60066 : 34..148(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1825..1939\n",
      "Associations = {\"note\"=>\"match: TC65864 : 34..148(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1825..1939\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Gossypium_raimondii-93 : 34..148(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1825..1939\n",
      "Associations = {\"note\"=>\"match: TA14308_29730 : 34..148(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 36..1979\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-171595 : 3..985(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 227..1978\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-171708 : 1..486(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 111..428\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-298790 : 1..319(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 448..1991\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Arabidopsis_thaliana-5380411836 : 2..584(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 34..2031\n",
      "Associations = {\"note\"=>\"match: At.287 : 1..1037(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 289..1948\n",
      "Associations = {\"note\"=>\"match: TC293195 : 104..802(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 62..1978\n",
      "Associations = {\"note\"=>\"match: TC301920 : 3..958(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1078..1957\n",
      "Associations = {\"note\"=>\"match: TC305197 : 244..677(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 227..1978\n",
      "Associations = {\"note\"=>\"match: TC305959 : 1..486(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1268..2031\n",
      "Associations = {\"note\"=>\"match: TC306275 : 1..496(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 34..1978\n",
      "Associations = {\"note\"=>\"match: TA29636_3702 : 2..985(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 227..1978\n",
      "Associations = {\"note\"=>\"match: TA29642_3702 : 1..486(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = AL132971.2:7054..8774\n",
      "Associations = {\"note\"=>\"match: TA29643_3702 : 2..627(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 133..327\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Brassica_napus-32240 : 60..256(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 133..1960\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Brassica_napus-56781 : 67..914(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 133..1960\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Brassica_napus-87128 : 29..871(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 137..1958\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Brassica_napus-120979 : 61..920(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 264..1958\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Brassica_napus-120980 : 50..738(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 133..327\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Brassica_napus-137858 : 52..248(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 137..782\n",
      "Associations = {\"note\"=>\"match: PUT-161a-Brassica_napus-10137959 : 13..598(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1264..1948\n",
      "Associations = {\"note\"=>\"match: Bna.7355 : 23..437(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 133..1804\n",
      "Associations = {\"note\"=>\"match: Bna.25043 : 41..731(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 137..1958\n",
      "Associations = {\"note\"=>\"match: TC20417 : 14..873(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 264..1958\n",
      "Associations = {\"note\"=>\"match: TC26043 : 14..702(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 133..1960\n",
      "Associations = {\"note\"=>\"match: TC58775 : 24..866(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 137..1954\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Brassica_rapa-31219 : 26..880(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 168..1839\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Brassica_rapa-25645021 : 13..699(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 289..474\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Brassica_rapa-25845014 : 1..162(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 137..1954\n",
      "Associations = {\"note\"=>\"match: Bra.647 : 26..879(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 289..474\n",
      "Associations = {\"note\"=>\"match: Bra.15464 : 1..162(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 76..1954\n",
      "Associations = {\"note\"=>\"match: Rra.2107 : 4..917(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 288..1960\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Raphanus_sativus-2148 : 26..714(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 76..1962\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Raphanus_sativus-22799 : 39..957(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 96..1093\n",
      "Associations = {\"note\"=>\"match: PUT-163a-Raphanus_sativus-22813 : 46..521(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 96..1852\n",
      "Associations = {\"note\"=>\"match: Rsa.8014 : 47..838(1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1769..1929\n",
      "Associations = {\"note\"=>\"match: PUT-159a-Cichorium_intybus-7959 : 139..299(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1842..1941\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Ipomoea_nil-21407 : 90..189(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1842..1941\n",
      "Associations = {\"note\"=>\"match: TA6369_35883 : 90..189(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = AL132971.2:9169..9654\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Coffea_canephora-13381 : 63..392(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1818..1948\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Mimulus_guttatus-7984 : 1..131(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1826..1918\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Mimulus_guttatus-8568 : 183..275(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1854..1960\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Mimulus_guttatus-419381 : 20..120(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1854..1960\n",
      "Associations = {\"note\"=>\"match: TA1209_4155 : 20..120(-1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1765..1941\n",
      "Associations = {\"note\"=>\"match: PUT-157a-Antirrhinum_majus-12340 : 92..267(- 1)\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1765..1941\n",
      "Associations = {\"note\"=>\"match: TA4478_4151 : 92..267(-1)\"}\n",
      "Bio::Feature\n",
      "repeat_region\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(823..829)\n",
      "Associations = {\"note\"=>\"dust repeat: matches 20119962..20119968(1) of consensus\"}\n",
      "Bio::Feature\n",
      "repeat_region\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(906..924)\n",
      "Associations = {\"note\"=>\"dust repeat: matches 20120045..20120063(1) of consensus\"}\n",
      "Bio::Feature\n",
      "repeat_region\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1399..1406)\n",
      "Associations = {\"note\"=>\"dust repeat: matches 20120538..20120545(1) of consensus\"}\n",
      "Bio::Feature\n",
      "repeat_region\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1558..1591)\n",
      "Associations = {\"note\"=>\"dust repeat: matches 20120697..20120730(1) of consensus\"}\n",
      "Bio::Feature\n",
      "repeat_region\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(1979..2063)\n",
      "Associations = {\"note\"=>\"dust repeat: matches 20121118..20121202(1) of consensus\"}\n",
      "Bio::Feature\n",
      "repeat_region\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = complement(2169..2175)\n",
      "Associations = {\"note\"=>\"dust repeat: matches 20121308..20121314(1) of consensus\"}\n",
      "Bio::Feature\n",
      "misc_feature\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "POSITION = 1..2175\n",
      "Associations = {\"note\"=>\"contig AL132971.2 7316..9490(1)\"}\n",
      "Bio::EMBL\n",
      "#  - \n"
     ]
    }
   ],
   "source": [
    "datafile2 = Bio::FlatFile.auto('At3g54340.embl')\n",
    "puts datafile2.class  \n",
    "\n",
    "datafile2.each_entry do |entry| # the FILE is not the same as the RECORD - multiple records can exist in a file\n",
    "# shows accession and organism\n",
    "  puts entry.class\n",
    "  puts \"# #{entry.accession} - #{entry.species}\"\n",
    "\n",
    "  # iterates over each element in 'features'\n",
    "  entry.features.each do |feature|\n",
    "    position = feature.position\n",
    "    puts \"\\n\\n\\n\\nPOSITION = #{position}\"\n",
    "    qual = feature.assoc            # feature.assoc gives you a hash of Bio::Feature::Qualifier objects \n",
    "                                    # i.e. qualifier['key'] = value  for example qualifier['gene'] = \"CYP450\")\n",
    "    puts \"Associations = #{qual}\"\n",
    "    # skips the entry if \"/translation=\" is not found\n",
    "    next unless qual['translation']    # this is an indication that the feature is a transcript\n",
    "\n",
    "    # collects gene name and so on and joins it into a string\n",
    "    gene_info = [\n",
    "      qual['gene'], qual['product'], qual['note'], qual['function']\n",
    "    ].compact.join(', ')\n",
    "    puts \"TRANSCRIPT FOUND!\\nGene Info:  #{gene_info}\"\n",
    "    # shows nucleic acid sequence\n",
    "    puts \"\\n\\n>NA splicing('#{position}') : #{gene_info}\"\n",
    "    puts entry.naseq.class   # this is a Bio::Sequence::NA    Look at the documentation to understand the .splicing() method\n",
    "    puts entry.naseq.splice(position)  # http://bioruby.org/rdoc/Bio/Sequence/Common.html#method-i-splice\n",
    "\n",
    "    # shows amino acid sequence translated from nucleic acid sequence\n",
    "    puts \"\\n\\n>AA translated by splicing('#{position}').translate\"\n",
    "    puts entry.naseq.splicing(position).translate\n",
    "\n",
    "    # shows amino acid sequence in the database entry (/translation=)\n",
    "    puts \"\\n\\n>AA original translation\"\n",
    "    puts qual['translation']\n",
    "  end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prove that you understand\n",
    "\n",
    "easy:  Modify the code above to print a list of the SNPs in this record.\n",
    "\n",
    "harder:  For each SNP, report if it is in an exon or an intron\n",
    "\n",
    "NOTE: from the Bio::Feature documentation\n",
    "\n",
    "\n",
    "### Attributes\n",
    "    feature[RW]\n",
    "    Returns type of feature in String (e.g 'CDS', 'gene')\n",
    "    \n",
    "    position[RW]\n",
    "    Returns position of the feature in String (e.g. 'complement(123..146)')\n",
    "\n",
    "    qualifiers[RW]\n",
    "    Returns an Array of Qualifier objects.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Alignments \n",
    "\n",
    "It is common to want to align sequences.  BioRuby can do this for you in simple cases:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a?gc?\n",
      "ahgcr\n"
     ]
    }
   ],
   "source": [
    "require 'bio'\n",
    "\n",
    "seqs = [ 'atgca', 'aagca', 'acgca', 'acgcg' ]\n",
    "seqs = seqs.collect{ |x| Bio::Sequence::NA.new(x) }   # look up the documentation for .collect.  How is it different from .each?\n",
    "# creates alignment object\n",
    "a = Bio::Alignment.new(seqs)\n",
    "puts a.consensus \n",
    "puts a.consensus_iupac"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "  \n",
    "</pre>\n",
    "Try adding a few nucleotides to one of the sequences.  What happens in the alignment?  Add the same nucleotides to the other sequences, one-by-one... what happens in the alignment?   What do you learn from this?\n",
    "\n",
    "\n",
    "### more complicated alignments, e.g. with gaps, requires an alignment algorithm.\n",
    "\n",
    "A common way to do alignments is ClustalW.  This is installed on your virtual machine.  \n",
    "\n",
    "To access ClustalW from BioRuby's Alignment object, you do this:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BioRuby simple alignment\n",
      "??g???t?\n",
      "mhgsddt?\n",
      "\n",
      "ClustalW alignment\n",
      "???G???T?\n",
      "?wvgyrkt?\n"
     ]
    }
   ],
   "source": [
    "require 'bio'\n",
    "\n",
    "seqs = [ 'ctggtatt', 'aagcatt', 'acgcagtt', 'acgcgtt' ]\n",
    "seqs = seqs.collect{ |x| Bio::Sequence::NA.new(x) }   # look up the documentation for .collect.  How is it different from .each?\n",
    "# creates alignment object\n",
    "a = Bio::Alignment.new(seqs)\n",
    "puts \"BioRuby simple alignment\"\n",
    "puts a.consensus_string\n",
    "puts a.consensus_iupac\n",
    "puts\n",
    "puts \"ClustalW alignment\"\n",
    "factory = Bio::ClustalW.new\n",
    "a2 = a.do_align(factory)\n",
    "puts a2.consensus_string\n",
    "puts a2.consensus_iupac"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.4.2",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
