{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=AT1g21400,AT5g19120,AT2g13360,AT4g05180,AT1g22690,AT2g45170,AT4g09650,AT5g55620,AT1g31330,AT1g80440,AT3g28270,AT4g12800,AT5g04140,AT5g64040,AT1g29660,AT1g15820,AT1g64500,AT1g03130,AT2g20670,AT4g17090\n",
      "AT1G21400\n",
      "AT5G19120\n",
      "AT2G13360\n",
      "AT4G05180\n",
      "AT1G22690\n",
      "AT1G22700\n",
      "AT2G45170\n",
      "AT4G09650\n",
      "AT5G55620\n",
      "AT1G31330\n",
      "AT1G80440\n",
      "AT3G28270\n",
      "AT4G12800\n",
      "AT5G04140\n",
      "AT5G04150\n",
      "AT5G64030\n",
      "AT5G64040\n",
      "AT1G29660\n",
      "AT1G15810\n",
      "AT1G15820\n",
      "AT1G05207\n",
      "AT1G64500\n",
      "AT1G03130\n",
      "AT1G03140\n",
      "AT2G20670\n",
      "AT4G17085\n",
      "AT4G17090\n",
      "AT4G06665\n",
      "DONE\n"
     ]
    }
   ],
   "source": [
    "require 'net/http' \n",
    "require 'bio'\n",
    "\n",
    "def fetch(uri_str)  # this \"fetch\" routine does some basic error-handling.  \n",
    "  address = URI(uri_str)  \n",
    "  response = Net::HTTP.get_response(address)\n",
    "  case response   # the \"case\" block allows you to test various conditions... it is like an \"if\", but cleaner!\n",
    "    when Net::HTTPSuccess then  # when response is of type Net::HTTPSuccess\n",
    "      return response  # return that response object\n",
    "    else\n",
    "      raise Exception, \"Something went wrong... the call to #{uri_str} failed; type #{response.class}\"\n",
    "      response = False\n",
    "      return response  # now we are returning False\n",
    "    end\n",
    "end\n",
    "  \n",
    "#puts Bio::GenBank::DELIMITER\n",
    "#abort \"stopping \"\n",
    "\n",
    "# open the gene list for input, open an EMBL file for output\n",
    "genes = File.open('./files/short_gene_list.txt', 'r')\n",
    "embloutput = File.open('./files/ARA.embl', 'w')\n",
    "\n",
    "genearray = genes.read.split() # this will read each line into an array\n",
    "geneids=genearray.join(\",\")  # join the elements of the array with a comma\n",
    "genes.close  # close the input file\n",
    "\n",
    "#  change back to EMBL format, get all of the geneids as one dump\n",
    "url = \"http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=#{geneids}\"\n",
    "  \n",
    "puts url\n",
    "res = fetch(url)  # we really should check that the return value is valid, but... \n",
    "embloutput.write(res.body)\n",
    "embloutput.close\n",
    "\n",
    "# now open the EMBL file for input, and open a FASTA file for output\n",
    "emblinput = File.open('./files/ARA.embl', 'r')\n",
    "fastaoutput = File.open('./files/ARA.fa', 'w')\n",
    "  \n",
    "  # Read all lines from the file split by the GenBank delimiter\n",
    "  while entry = emblinput.gets(Bio::EMBL::DELIMITER)   # this is the first time you have seen a constant\n",
    "    record = Bio::EMBL.new(entry)      # creates GenBank object\n",
    "\n",
    "    record.features do |feature|\n",
    "      next unless feature.feature == \"gene\"\n",
    "      puts feature.assoc[\"gene\"]\n",
    "      fastaoutput.puts(\">#{feature.assoc[\"gene\"]}\")         # Accession\n",
    "      fastaoutput.puts(record.naseq)                    # Nucleic acid sequence \n",
    "\n",
    "    end\n",
    "  end\n",
    "  puts \"DONE\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## now look at your FASTA file again - Much better!\n",
    "\n",
    "You can now index that FASTA file for BLAST exactly as you did with the Ebola\n",
    "\n",
    "When you search that database, the <code>hit.hit_id</code> will give you the AGI Locus Code, instead of \"4\", \"5\", ....\n",
    "\n",
    "\n",
    "Please do this now, and make sure it is working correctly."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.4.2",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
